### PR TITLE
Adding compatible platforms to service spec

### PR DESCRIPTION
**Pull request by nishanttotla**
Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

This is a follow-up to #32388 (Related to: #31348).

It extracts platform information that is already present after the call to the `/distribution/json` endpoint (#32061), and adds it to the service spec. This platform information is used as a Swarm scheduling filter, to prevent tasks for running on nodes where they shouldn't run.

CLI changes in https://github.com/docker/cli/pull/30 need to be merged before this actually is enabled end to end. We'll also have to merge https://github.com/moby/moby/pull/33228, which includes a SwarmKit bug fix that is required for smooth functioning of this feature.

### CODE REVIEW

File: client/service_create.go [hunk 1]
```diff
@@ -71,6 +74,19 @@ func imageWithDigestString(image string, dgst digest.Digest) string {
 	return ""
 }
 
+// updateServicePlatforms
+func updateServicePlatforms(placement *swarm.Placement, distributionInspect registrytypes.DistributionInspect) {
+	if placement == nil {
+		placement = &swarm.Placement{}
```
Comment (aaronlehmann) [inline client/service_create.go]: This is changing the pointer, not the value

File: client/service_create.go [hunk 2]
```diff
@@ -71,6 +74,19 @@ func imageWithDigestString(image string, dgst digest.Digest) string {
 	return ""
 }
 
+// updateServicePlatforms
```
Comment (aaronlehmann) [inline client/service_create.go]: ?

**Comment by aaronlehmann** [comment]
Tests?

**Comment by nishanttotla** [comment]
@aaronlehmann I think my commit was overwritten, so this is stale. Updating it.

Where should the tests ideally be put for this?

**Comment by aaronlehmann** [comment]
> Where should the tests ideally be put for this?

client/service_create_test.go

**Comment by nishanttotla** [comment]
@aaronlehmann updated PR and added a unit test (the unit test is somewhat of a hack to test the platform response at the API level, but hopefully it is okay).

**Comment by aaronlehmann** [comment]
Thanks. Could you please extend the unit test to check that the correct platform information is being passed to the `/services/create` endpoint?

**Comment by nishanttotla** [comment]
@aaronlehmann done.

**Comment by aaronlehmann** [comment]
I'm not seeing it. The request body to `/services/create` is never unmarshalled.

File: client/service_create_test.go [hunk 3]
```diff
@@ -55,3 +58,66 @@ func TestServiceCreate(t *testing.T) {
...
+
+	client := &Client{
+		client: newMockClient(func(req *http.Request) (*http.Response, error) {
+			if strings.HasPrefix(req.URL.Path, "/services/create") {
+				// check if the /distribution endpoint returned correct output
+				err := json.NewDecoder(distributionInspectBody).Decode(&distributionInspect)
+				if err != nil {
+					return nil, err
+				}
+				if distributionInspect.Platforms != platforms {
```
Comment (aaronlehmann) [inline client/service_create_test.go]: This comparison won't work. They are both slices.

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM, not sure if we should sort the platforms array but shouldn't be a blocker

**Comment by tiborvass** [comment]
Should be fine like this, sorting is usually for display

**Comment by nishanttotla** [comment]
@tiborvass we could do it as a follow-up, it's not critical right now.

**Pull request merged**